### PR TITLE
Initial release

### PR DESCRIPTION
**Pull request by moattarwork**
# Pull Request Template

The PR contains the initial release of the Export functionality for Easify

New Features

- Export data as Csv to S3, Azure Blob etc

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

It has been tested through unit and integration tests

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] My changes fulfill the acceptance criteria for the story or bug
- [ ] I made/planned the required configuration changes on CI/CD

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,30 +1,288 @@
...
+It also send the notifications on success and error result to an Scheduler (which has been implemented the required interface).
 
-Description of how to start working with library
+![Architecture](docs/images/architecture.png)
 
-### How to use
+## Common Layer
 
-The usage of the library or link to the relevant wiki page need to be here.
+The client library provides the common structure which is shared between Domain & Scheduler API's. It contains Common and Client libraries.
```
Comment (davidcassell) [inline README.md:16]: Apostrophe

File: README.md [hunk 2]
```diff
@@ -1,30 +1,288 @@
-Library Name 
-============
+# Easify Exports
 
-Brief description of the component or library.
+The repo contains a series of Easify componenets which are providing the export functionality of any data source to CsV files targeted to
```
Comment (davidcassell) [inline README.md:3]: Casing on Csv

File: README.md [hunk 3]
```diff
@@ -1,30 +1,288 @@
...
+## Common Layer
 
-The usage of the library or link to the relevant wiki page need to be here.
+The client library provides the common structure which is shared between Domain & Scheduler API's. It contains Common and Client libraries.
 
-### How to Engage, Contribute, and Give Feedback
+### Client
 
-Description of the steps or process to be a contributor to the project.
+The client library contains all the registration and client access library for the scheduler to talk to Exporters in different domain API.
```
Comment (davidcassell) [inline README.md:20]: "call Exporters" vs "talk to Exporters"

File: src/Easify.Exports.Agent.UnitTests/CsvStorageExporterTests.cs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+                ILogger<CsvStorageExporter<Sample>> logger) : base(fileExporter, notifierBuilder, logger)
+            {
+                _dataProvider = dataProvider;
+            }
+
+            protected override string ExportFilePrefix => "Sample";
+
+            protected override async Task<IEnumerable<Sample>> PrepareDataAsync(ExportExecutionContext executionContext)
+            {
+                return _dataProvider();
```
Comment (davidcassell) [inline src/Easify.Exports.Agent.UnitTests/CsvStorageExporterTests.cs]: Clean up the Github action warnings? (throughout)

**Review by davidcassell — APPROVED** [review]
A couple of comments for you to look at, but approved anyway

File: README.md [hunk 5]
```diff
@@ -1,30 +1,288 @@
...
-Brief description of the component or library.
+The repo contains a series of Easify componenets which are providing the export functionality of any data source to CsV files targeted to 
+  - S3, 
+  - Azure Blob
+  - File Share 
+  - In Memory file store
+  - Or a combination of the targets 
 
-## Get Started
+It also send the notifications on success and error result to an Scheduler (which has been implemented the required interface).
```
Comment (french-justin) [inline README.md:10]: It also sends the notifications on success and error to a Scheduler (which has implemented the required interface)

File: README.md [hunk 6]
```diff
@@ -1,30 +1,288 @@
...
-* [Roadmap](./docs/ROADMAP.md): The schedule and milestone themes for project.
+or
 
-## Reporting bugs and useful features
+```cmd
+Install-Package Easify.Exports.Client
+```
 
-Security issues and bugs should be reported by creating the relevant features and bugs in issues sections
+The **AddExporterBuilder** extension method will help on registering the required exporters
```
Comment (french-justin) [inline README.md]: will help with registering

File: src/Easify.Exports.Client/ExporterClientBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        public IExporterClient Build(string name)
+        {
+            if (!_exporterCache.ContainsKey(name))
+                throw new ExporterNotFoundException(
+                    $"No exporter found with name: {name}. The {name} should be added using AddClient in configuration");
+
+            return RestClient.For<IExporterClient>(_exporterCache[name]);
+
+            // TODO: Need to be replaced with this due to the issue
+            // return _clientBuilder.Build<IExporterClient>(_exporterCache[name]);
```
Comment (french-justin) [inline src/Easify.Exports.Client/ExporterClientBuilder.cs:38]: Is this TODO still valid? If so what issue?

File: src/Easify.Exports.IntegrationTests/CsvFileGenerationWithServiceCollectionTests.cs [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+            actual.HasError.Should().BeTrue();
+            actual.RecordCount.Should().Be(0);
+            actual.Error.Should()
+                .Be(
+                    $"Error in creating export configuration '{typeof(SampleEntity2)}'. Make sure the ClassMap has been registered");
+        }
+
+        // [Fact]
+        // TODO: Should be revised
+        public async Task Should_ExportAsync_UploadTheFileInAws()
```
Comment (french-justin) [inline src/Easify.Exports.IntegrationTests/CsvFileGenerationWithServiceCollectionTests.cs:135]: Are you fixing these?

File: src/Easify.Exports.IntegrationTests/CsvFileGenerationWithServiceCollectionTests.cs [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+            var sut = serviceProvider.GetRequiredService<IFileExporter>();
+
+            var actual = await sut.ExportAsync(entities, new ExporterOptions(DateTime.Today, storageTargets));
+
+            actual.Should().NotBe(null);
+            actual.HasError.Should().BeFalse();
+        }
+
+        //[Fact]
+        public async Task Should_ExportAsync_UploadTheFileInAzureBlobStorageWithEmulator()
```
Comment (french-justin) [inline src/Easify.Exports.IntegrationTests/CsvFileGenerationWithServiceCollectionTests.cs:220]: Same as above

File: src/Easify.Exports.UnitTests/ServiceCollectionExtensionsTests.cs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            services.AddLocalDiskStorage();
+
+            var provider = services.BuildServiceProvider();
+
+            // ACT
+            var sut = provider.GetRequiredService<ICsvStorageTargetResolver>();
+
+            // ASSERT
+            sut.Resolve(StorageTargetType.LocalDisk).Should().NotBeNull();
+
```
Comment (french-justin) [inline src/Easify.Exports.UnitTests/ServiceCollectionExtensionsTests.cs]: JL

File: src/Easify.Exports.UnitTests/ServiceCollectionExtensionsTests.cs [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+            services.AddInMemoryStorage();
+
+            var provider = services.BuildServiceProvider();
+
+            // ACT
+            var sut = provider.GetRequiredService<ICsvStorageTargetResolver>();
+
+            // ASSERT
+            sut.Resolve(StorageTargetType.InMemory).Should().NotBeNull();
+
```
Comment (french-justin) [inline src/Easify.Exports.UnitTests/ServiceCollectionExtensionsTests.cs]: JL

File: src/Easify.Exports/Csv/CsvExportConfiguration.cs [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+using Easify.Exports.Storage;
+
+namespace Easify.Exports.Csv
+{
+    public class CsvExportConfiguration
+    {
+        public StorageTarget[] Targets { get; set; } = { };
+        public string FileName { get; set; }
+        public CsvConfiguration Configuration { get; set; }
+
```
Comment (french-justin) [inline src/Easify.Exports/Csv/CsvExportConfiguration.cs]: JL

**Review by french-justin — CHANGES_REQUESTED** [review]
Looks good, just some changes to look at

File: src/Easify.Exports/Csv/CsvExportConfiguration.cs (hunk 12 above)
Comment (moattarwork) [reply src/Easify.Exports/Csv/CsvExportConfiguration.cs]: 😄

File: src/Easify.Exports.UnitTests/ServiceCollectionExtensionsTests.cs (hunk 10 above)
Comment (moattarwork) [reply src/Easify.Exports.UnitTests/ServiceCollectionExtensionsTests.cs]: :)

File: src/Easify.Exports.IntegrationTests/CsvFileGenerationWithServiceCollectionTests.cs (hunk 9 above)
Comment (moattarwork) [reply src/Easify.Exports.IntegrationTests/CsvFileGenerationWithServiceCollectionTests.cs:220]: Cannot be running with no access to AWS. Needs to find another way

File: src/Easify.Exports.IntegrationTests/CsvFileGenerationWithServiceCollectionTests.cs (hunk 8 above)
Comment (moattarwork) [reply src/Easify.Exports.IntegrationTests/CsvFileGenerationWithServiceCollectionTests.cs:135]: Same as below

File: src/Easify.Exports.Client/ExporterClientBuilder.cs (hunk 7 above)
Comment (moattarwork) [reply src/Easify.Exports.Client/ExporterClientBuilder.cs:38]: Not at the moment but sometime in the future it should be. It need some contextual info

**Review by french-justin — APPROVED** [review]

**Pull request merged**